native/core/src/store/index.rs

```diff
@@ -0,0 +1,293 @@
+/// latest first
+pub struct LifoIndex<T>
+where
+    T: 'static + Clone + Eq,
+{
+    vector: ObservableVector<T>,
+}
+
+impl<T> LifoIndex<T>
+where
+    T: 'static + Clone + Eq,
+{
+    pub fn new_with(value: T) -> Self {
+        let mut m = LifoIndex::default();
+        m.insert(value);
+        m
+    }
+    /// Insert the element at the front
+    pub fn insert(&mut self, value: T) {
+        self.vector.push_front(value);
+    }
+
+    /// All instances of this element from the vector
+    pub fn remove(&mut self, value: &T) {
+        let mut t = self.vector.transaction();
+        let mut entries = t.entries();
+        while let Some(entry) = entries.next() {
+            if &*entry == value {
+                ObservableVectorTransactionEntry::remove(entry);
+            }
+        }
+        t.commit();
+    }
+
+    /// Returns the current list of values in order of when they were added
+    pub fn values(&self) -> Vec<&T> {
+        self.vector.iter().collect()
+    }
+
+    pub fn update_stream(&self) -> impl Stream<Item = VectorDiff<T>> {
+        self.vector.subscribe().into_stream()
+    }
+}
+
+impl<T> Default for LifoIndex<T>
+where
+    T: 'static + Clone + Eq,
+{
+    fn default() -> Self {
+        Self {
+            vector: Default::default(),
+        }
+    }
+}
+
+impl<T> Deref for LifoIndex<T>
+where
+    T: 'static + Clone + Eq,
+{
+    type Target = ObservableVector<T>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.vector
+    }
+}
+
+pub enum StoreIndex {
+    Lifo(LifoIndex<OwnedEventId>),
+    Ranked(RankedIndex<MilliSecondsSinceUnixEpoch, OwnedEventId>),
+}
+
+impl StoreIndex {
+    pub fn new_for(key: &IndexKey, meta: &EventMeta) -> StoreIndex {
+        match key {
+            IndexKey::ObjectHistory(_) | IndexKey::RoomHistory(_) => StoreIndex::Ranked(
+                RankedIndex::new_with(meta.origin_server_ts, meta.event_id.clone()),
+            ),
+            _ => StoreIndex::Lifo(LifoIndex::new_with(meta.event_id.clone())),
+        }
+    }
+
+    pub fn insert(&mut self, meta: &EventMeta) {
+        match self {
+            StoreIndex::Lifo(l) => l.insert(meta.event_id.clone()),
+            StoreIndex::Ranked(r) => r.insert(meta.origin_server_ts, meta.event_id.clone()),
+        }
+    }
+
+    /// All instances of this element from the vector
+    pub fn remove(&mut self, value: &OwnedEventId) {
+        match self {
+            StoreIndex::Lifo(lifo_index) => lifo_index.remove(value),
+            StoreIndex::Ranked(ranked_index) => ranked_index.remove(value),
+        }
+    }
+
+    /// Returns the current list of values in order of when they were added
+    pub fn values(&self) -> Vec<&OwnedEventId> {
+        match self {
+            StoreIndex::Lifo(lifo_index) => lifo_index.values(),
+            StoreIndex::Ranked(ranked_index) => ranked_index.values(),
+        }
+    }
+
+    // pub fn update_stream(&self) -> impl Stream<Item = VectorDiff<OwnedEventId>> {
```

Comment:
this is the future API that _should_ work, but the rust type system currently doesn't allow us to do this way. I kept it here for reference, but it will be changed in the future version soon.